### PR TITLE
MC-179 : Add Delta product export

### PR DESCRIPTION
**Pull request by willy-ahva**
Bug fix: n
Feature addition: y
Backwards compatibility break: n
Phpspec specification passes: y
Checkstyle issues: n
ChangeLog updated: n
Fixes the following issue: MC-179

### CODE REVIEW

File: Entity/DeltaProductExport.php [hunk 1]
```diff
@@ -31,13 +35,13 @@ public function getId()
...
 
     /**
-     * @param \DateTime $date
+     * @param \DateTime $lastExport
      *
-     * @return ProductExport
+     * @return DeltaProductExport
      */
-    public function setDate($date)
+    public function setLastExport($lastExport)
```
Comment (fitn) [inline Entity/DeltaProductExport.php]: To test with \DateTime typehint

File: Manager/ProductExportManager.php [hunk 2]
```diff
@@ -16,104 +19,108 @@
...
-                ';
-            }
+        $conn = $this->entityManager->getConnection();
 
-            $q = $conn->prepare($sql);
-            $date = $now->format('Y-m-d H:i:s');
-            $productId = $product->getId();
-            $jobInstanceId = $jobInstance->getId();
+        $jobInstance->getId();
+        $product->getId();
```
Comment (fitn) [inline Manager/ProductExportManager.php]: 2 lines useless

File: Manager/ProductExportManager.php [hunk 3]
```diff
@@ -16,104 +19,108 @@
...
+                UPDATE $table
+                SET last_export = :last_export
+                WHERE product_id = :product_id AND job_instance_id = :job_instance_id
+            ";
         }
+
+        $query = $conn->prepare($sql);
+        $lastExport = $now->format('Y-m-d H:i:s');
+        $productId = $product->getId();
+        $jobInstanceId = $jobInstance->getId();
```
Comment (fitn) [inline Manager/ProductExportManager.php]: You can avoid to store in variable and directly use it in the bindParam ;)

File: Manager/ProductExportManager.php [hunk 4]
```diff
@@ -16,104 +19,108 @@
...
+
+        $query->bindParam(':last_export', $lastExport, PDO::PARAM_STR);
+        $query->bindParam(':product_id', $productId, PDO::PARAM_INT);
+        $query->bindParam(':job_instance_id', $jobInstanceId, PDO::PARAM_INT);
```
Comment (fitn) [inline Manager/ProductExportManager.php]: $now->format('Y-m-d H:i:s');
$product->getId();
$jobInstance->getId();

File: Manager/ProductExportManager.php [hunk 5]
```diff
@@ -16,104 +19,108 @@
...
      * @param JobInstance $jobInstance
      */
-    public function updateProductExports($products, JobInstance $jobInstance)
+    public function updateProductExport($identifier, JobInstance $jobInstance)
     {
-        foreach ($products as $product) {
-            $this->updateProductExport($product->getIdentifier(), $jobInstance);
+        $product = $this->productRepository->findByReference((string) $identifier);
+
+        if (null != $product) {
```
Comment (fitn) [inline Manager/ProductExportManager.php]: !==

File: Manager/ProductExportManager.php (hunk 3 above)
Comment (fitn) [inline Manager/ProductExportManager.php]: You can just instanciate $now here ;)

File: Resources/config/batch_jobs.yml [hunk 6]
```diff
@@ -53,37 +53,16 @@ connector:
                 export_simple_products:
                     title:     pim_magento_connector.steps.simple_product_export.export.title
                     services:
-                        reader:    pim_magento_connector.reader.doctrine.product
+                        reader:    pim_magento_connector.reader.orm.product_delta
```
Comment (fitn) [inline Resources/config/batch_jobs.yml]: :+1:

Comment (fitn) [inline Resources/config/batch_jobs.yml]: pim_magento_connector.reader.product_delta
There is only one reader in the DI but it will work on MongoDB OR on ORM ;)

File: Resources/config/batch_jobs.yml [hunk 7]
```diff
@@ -53,37 +53,16 @@ connector:
...
                 prune_simple_products:
                     title: pim_magento_connector.steps.simple_product_clean.clean.title
                     class: "%pim_magento_connector.step.prune.class%"
                     services:
                         cleaner: pim_magento_connector.cleaner.product
-                prune_configurable_products:
-                    title: pim_magento_connector.steps.configurable_product_clean.clean.title
-                    class: "%pim_magento_connector.step.prune.class%"
-                    services:
-                        cleaner: pim_magento_connector.cleaner.configurable
```
Comment (fitn) [inline Resources/config/batch_jobs.yml]: Why did you remove these steps? It should not be removed, it should be fixed :)

File: Resources/config/entities.yml [hunk 8]
```diff
@@ -1,8 +1,8 @@
...
-    pim_magento_connector.entity.simple_mapping.class:    Pim\Bundle\MagentoConnectorBundle\Entity\SimpleMapping
-    pim_magento_connector.entity.product_export.class:    Pim\Bundle\MagentoConnectorBundle\Entity\ProductExport
-    pim_magento_connector.entity.simple_mapping.class:    Pim\Bundle\MagentoConnectorBundle\Entity\SimpleMapping
+    pim_magento_connector.entity.category_mapping.class:                 Pim\Bundle\MagentoConnectorBundle\Entity\MagentoCategoryMapping
+    pim_magento_connector.entity.family_mapping.class:                   Pim\Bundle\MagentoConnectorBundle\Entity\MagentoFamilyMapping
+    pim_magento_connector.entity.group_mapping.class:                    Pim\Bundle\MagentoConnectorBundle\Entity\MagentoGroupMapping
+    pim_magento_connector.entity.attribute_mapping.class:                Pim\Bundle\MagentoConnectorBundle\Entity\MagentoAttributeMapping
+    pim_magento_connector.entity.simple_mapping.class:                   Pim\Bundle\MagentoConnectorBundle\Entity\SimpleMapping
+    pim_magento_connector.entity.simple_mapping.class:                   Pim\Bundle\MagentoConnectorBundle\Entity\SimpleMapping
```
Comment (fitn) [inline Resources/config/entities.yml]: duplicate line

File: Resources/config/managers.yml [hunk 9]
```diff
@@ -1,17 +1,17 @@
...
+    pim_magento_connector.manager.group_mapping.class:       Pim\Bundle\MagentoConnectorBundle\Manager\AttributeGroupMappingManager
+    pim_magento_connector.manager.attribute_mapping.class:   Pim\Bundle\MagentoConnectorBundle\Manager\AttributeMappingManager
+    pim_magento_connector.manager.simple_mapping.class:      Pim\Bundle\MagentoConnectorBundle\Manager\SimpleMappingManager
+    pim_magento_connector.manager.association_type.class:    Pim\Bundle\MagentoConnectorBundle\Manager\AssociationTypeManager
+    pim_magento_connector.manager.product_value.class:       Pim\Bundle\MagentoConnectorBundle\Manager\ProductValueManager
+    pim_magento_connector.manager.attribute.class:           Pim\Bundle\MagentoConnectorBundle\Manager\AttributeManager
+    pim_magento_connector.manager.group.class:               Pim\Bundle\MagentoConnectorBundle\Manager\GroupManager
+    pim_magento_connector.manager.product_export.class:      Pim\Bundle\MagentoConnectorBundle\Manager\ProductExportManager
+    pim_magento_connector.manager.simple_mapping.class:      Pim\Bundle\MagentoConnectorBundle\Manager\SimpleMappingManager
```
Comment (fitn) [inline Resources/config/managers.yml]: duplicate line

File: Resources/config/readers.yml [hunk 10]
```diff
@@ -47,22 +46,6 @@ services:
...
 
-    pim_magento_connector.reader.doctrine.product:
-        class: %pim_magento_connector.reader.doctrine.product.class%
-        parent: pim_base_connector.reader.doctrine.obsoleteproduct
-        arguments:
-            - @pim_magento_connector.manager.product_export
-
-    pim_magento_connector.reader.doctrine.product_value:
-        class: %pim_magento_connector.reader.doctrine.product.class%
+    pim_magento_connector.reader.orm.product_delta:
```
Comment (fitn) [inline Resources/config/readers.yml]: - orm. to be compliant with MongoDB

File: Resources/config/repositories.yml [hunk 11]
```diff
@@ -1,7 +1,8 @@
 parameters:
-    pim_magento_connector.repository.category.class: Pim\Bundle\MagentoConnectorBundle\Entity\Repository\CategoryRepository
-    pim_magento_connector.repository.currency.class: Pim\Bundle\MagentoConnectorBundle\Entity\Repository\CurrencyRepository
-    pim_magento_connector.repository.group.class:    Pim\Bundle\MagentoConnectorBundle\Entity\Repository\GroupRepository
+    pim_magento_connector.repository.category.class:      Pim\Bundle\MagentoConnectorBundle\Entity\Repository\CategoryRepository
+    pim_magento_connector.repository.currency.class:      Pim\Bundle\MagentoConnectorBundle\Entity\Repository\CurrencyRepository
+    pim_magento_connector.repository.group.class:         Pim\Bundle\MagentoConnectorBundle\Entity\Repository\GroupRepository
+    pim_magento_connector.repository.delta_product.class: Pim\Bundle\MagentoConnectorBundle\Entity\Repository\DeltaProductExportRepository
```
Comment (fitn) [inline Resources/config/repositories.yml]: delta_product_export

File: Resources/config/repositories.yml [hunk 12]
```diff
@@ -10,11 +11,17 @@ services:
...
 
-    pim_magento_connector.repository.currency:
-        class: %pim_magento_connector.repository.currency.class%
+    pim_magento_connector.repository.category:
+        class: %pim_magento_connector.repository.category.class%
+        arguments:
+            - '@doctrine.orm.entity_manager'
+            - '@pim_magento_connector.metadata.category'
+
+    pim_magento_connector.repository.delta_product:
```
Comment (fitn) [inline Resources/config/repositories.yml]: delta_product_export

File: Resources/config/batch_jobs.yml (hunk 7 above)
Comment (damien-carcel) [inline Resources/config/batch_jobs.yml]: Willy wanted to add it in the next PR, for better reading

Comment (willy-ahva) [inline Resources/config/batch_jobs.yml]: Because we are in the product delta export PR, not in the configurable delta export. This step will coming with configurable PR and will be fixed.

File: Manager/ProductExportManager.php [hunk 13]
```diff
@@ -16,104 +19,101 @@
...
+     *
      * @param string      $identifier
      * @param JobInstance $jobInstance
      */
     public function updateProductExport($identifier, JobInstance $jobInstance)
     {
-        $now = new \DateTime('now', new \DateTimeZone('UTC'));
         $product = $this->productRepository->findByReference((string) $identifier);
 
         if (null != $product) {
```
Comment (willy-ahva) [inline Manager/ProductExportManager.php]: !==

File: Manager/ProductExportManager.php [hunk 14]
```diff
@@ -16,104 +19,101 @@
...
-        }
-    }
 
     /**
      * Update product export date for the given product
+     *
      * @param string      $identifier
      * @param JobInstance $jobInstance
      */
     public function updateProductExport($identifier, JobInstance $jobInstance)
```
Comment (willy-ahva) [inline Manager/ProductExportManager.php]: type hinted before non type hinted

File: Manager/ProductExportManager.php [hunk 15]
```diff
@@ -16,104 +19,101 @@
...
-                    WHERE product_id = :product_id AND job_instance_id = :job_instance_id
-                ';
-            }
-
-            $q = $conn->prepare($sql);
-            $date = $now->format('Y-m-d H:i:s');
-            $productId = $product->getId();
-            $jobInstanceId = $jobInstance->getId();
+            $productExport = $this->productExportRepository->findOneBy(
+                array(
```
Comment (willy-ahva) [inline Manager/ProductExportManager.php]: [] instead of array()

File: Resources/config/managers.yml [hunk 16]
```diff
@@ -1,17 +1,16 @@
...
+    pim_magento_connector.manager.family_mapping.class:      Pim\Bundle\MagentoConnectorBundle\Manager\FamilyMappingManager
+    pim_magento_connector.manager.group_mapping.class:       Pim\Bundle\MagentoConnectorBundle\Manager\AttributeGroupMappingManager
+    pim_magento_connector.manager.attribute_mapping.class:   Pim\Bundle\MagentoConnectorBundle\Manager\AttributeMappingManager
+    pim_magento_connector.manager.simple_mapping.class:      Pim\Bundle\MagentoConnectorBundle\Manager\SimpleMappingManager
+    pim_magento_connector.manager.association_type.class:    Pim\Bundle\MagentoConnectorBundle\Manager\AssociationTypeManager
+    pim_magento_connector.manager.product_value.class:       Pim\Bundle\MagentoConnectorBundle\Manager\ProductValueManager
+    pim_magento_connector.manager.attribute.class:           Pim\Bundle\MagentoConnectorBundle\Manager\AttributeManager
+    pim_magento_connector.manager.group.class:               Pim\Bundle\MagentoConnectorBundle\Manager\GroupManager
+    pim_magento_connector.manager.product_export.class:      Pim\Bundle\MagentoConnectorBundle\Manager\ProductExportManager
```
Comment (willy-ahva) [inline Resources/config/managers.yml]: extra spaces to align parameters

File: Resources/config/managers.yml [hunk 17]
```diff
@@ -84,9 +83,9 @@ services:
     pim_magento_connector.manager.product_export:
```
Comment (willy-ahva) [inline Resources/config/managers.yml]: delta_product_export

File: Resources/config/managers.yml (hunk 16 above)
Comment (willy-ahva) [inline Resources/config/managers.yml]: delta_product_export

File: Manager/ProductExportManager.php [hunk 18]
```diff
@@ -16,104 +19,101 @@
  */
 class ProductExportManager
```
Comment (willy-ahva) [inline Manager/ProductExportManager.php]: DeltaProductExport

File: Resources/config/writers.yml [hunk 19]
```diff
@@ -18,11 +19,9 @@ services:
 
     pim_magento_connector.writer.product_delta:
         class: %pim_magento_connector.writer.product_delta.class%
+        parent: pim_magento_connector.writer.product
         arguments:
-            - '@pim_magento_connector.guesser.magento_webservice'
-            - '@pim_catalog.manager.channel'
             - '@pim_magento_connector.manager.product_export'
```
Comment (willy-ahva) [inline Resources/config/writers.yml]: delta_product_export

File: Manager/ProductExportManager.php [hunk 20]
```diff
@@ -16,104 +19,101 @@
...
-            $q->execute();
+    /**
+     * Update export date for the given product
+     *
+     * @param string           $export
+     * @param string           $table
+     * @param ProductInterface $product
+     * @param JobInstance      $jobInstance
+     */
+    protected function updateExport($export, $table, ProductInterface $product, JobInstance $jobInstance)
```
Comment (willy-ahva) [inline Manager/ProductExportManager.php]: type hinted before non type hinted

File: Manager/ProductExportManager.php (hunk 18 above)
Comment (fitn) [inline Manager/ProductExportManager.php]: Yep :+1:

File: Manager/ProductExportManager.php (hunk 13 above)
Comment (fitn) [inline Manager/ProductExportManager.php]: if ($product)

File: Manager/ProductExportManager.php (hunk 20 above)
Comment (fitn) [inline Manager/ProductExportManager.php]: Export should be typehinted

File: Resources/config/batch_jobs.yml (hunk 7 above)
Comment (fitn) [inline Resources/config/batch_jobs.yml]: Keep the steps like they were but don't remove them.

File: Manager/DeltaProductExportManager.php [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+                UPDATE $table
+                SET last_export = :last_export
+                WHERE product_id = :product_id AND job_instance_id = :job_instance_id
+            ";
+        }
+
+        $now           = new \DateTime('now', new \DateTimeZone('UTC'));
+        $formattedNow  = $now->format('Y-m-d H:i:s');
+        $productId     = $product->getId();
+        $jobInstanceId = $jobInstance->getId();
```
Comment (willy-ahva) [inline Manager/DeltaProductExportManager.php]: I re-put the here because it raises PHP notice `Only variables should be passed by reference`.

Comment (fitn) [inline Manager/DeltaProductExportManager.php]: okip :+1:

**Comment by fitn** [comment]
GTM
